### PR TITLE
Annotate argument in axes class match upstream

### PR DESCRIPTION
**Pull request by ksunden**
## PR Summary
`Annotate.__init__` changed it's signature in #13128, however it did not
change the python signature of the downstream functions which inherit its docstring, leading to a mismatch between the docstring and the signature.

I wasn't sure which version should have been put in the decorator, so I used the same as the parent, though I would expect to add the minor version, I guess.

Note: this does change pyplot without having a decorator to catch users who pass in the string explicitly by "s"... Not sure how to handle that

There was some debate in https://github.com/matplotlib/matplotlib/pull/12383#issuecomment-426872553 and #13128 regarding this variable change, that did mention the effects in pyplot, though seemed to indicate that it was acceptable.

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [x] New features are documented, with examples if plot related
- [x] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [x] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -771,9 +771,10 @@ def text(self, x, y, s, fontdict=None, withdash=False, **kwargs):
         self._add_text(t)
         return t
 
+    @cbook._rename_parameter("3.1", "s", "text")
```
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]:
```suggestion
    @cbook._rename_parameter("3.2", "s", "text")
```

**Comment by tacaswell** [comment]
Good catch!

We should have a bit of a think about what to do about pyplot.annoate though...

**Comment by anntzer** [comment]
there's no problem with pyplot.annotate?  it'll be subject to the same deprecation period.  it's _make_keyword_only that interacts poorly with pyplot.

**Comment by ksunden** [comment]
```
>>> plt.annotate(s="hi", xy=(0,0), xytext=(1,1))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: annotate() missing 1 required positional argument: 'text'
>>> plt.gca().annotate(s="hi", xy=(0,0), xytext=(1,1))
__main__:1: MatplotlibDeprecationWarning: The 's' parameter of annotate() has been renamed 'text' since Matplotlib 3.1; support for the old name will be dropped in 3.3.
Text(1, 1, 'hi')
```

With this patch applied, passing an explicit `s` kwarg to `pyplot.annotate` does cause a `TypeError`.
The function is not decorated with the `_rename_parameter` decorator.

**Comment by anntzer** [comment]
good catch :/
I guess this is another argument in favor of https://github.com/matplotlib/matplotlib/pull/14130#issuecomment-489461118 (@tacaswell do you want to comment on that?)

if the general problem doesn't get fixed quickly you can always go with manually writing the pyplot wrapper as a stopgap (https://github.com/matplotlib/matplotlib/pull/14130#issuecomment-493796194).

**Comment by ksunden** [comment]
I was just reminded of this, I've seen some PRs in recent days/weeks regarding annotate and some of the renaming of params and such, but did not track them closely enough to see if they were fixing this problem in another way, enabling this PR to work as intended, or just actually unrelated to this change entirely.

So, just checking in, seeing if there is anything that needs to be done to this PR to catch up (it appears to me that the docstring and the method signature are still mismatched as of now)

**Comment by jklymak** [comment]
There are no merge conflicts, so thats usually a good sign that no one has touched the section of the code you have touched.  OTOH, you have some documentation failures, and should rebase with master to make sure all the CI is using recent CI.

**Comment by anntzer** [comment]
Technically this breaks `plt.annotate(s=...)` without a deprecation, and the fix would be to merge https://github.com/matplotlib/matplotlib/pull/15254.  If we decide to ignore the APi break, I agree (some version of) this should go in ASAP.

**Comment by ksunden** [comment]
CI Failure looks unrelated to me... (Looks like two image files just failed to write properly, nothing to do with the actual code difference here)

I did test out #15254, the change proposed here is still needed (to Axes.annotate), but that PR does prevent the API deprecation problem. (There is a small merge conflict, as both edit pyplot.py, but it is only in the generated portion of that file, anyway)

**Comment by ksunden** [comment]
Closing and opening to rerun CI.

**Comment by ksunden** [comment]
Rebased on top of master after #15254 was merged

File: doc/api/next_api_changes/2019-08-12-KS.rst [hunk 2]
```diff
@@ -0,0 +1,9 @@
+Changes in parameter names
+``````````````````````````
+
+The parameter ``s`` to `.Axes.annotate` renamed to ``text``, matching
+`.Annotation`.
+This change propegates to ``pyplot.annotate``.
```
Comment (QuLogic) [inline doc/api/next_api_changes/2019-08-12-KS.rst]:
```suggestion
This change propagates to ``pyplot.annotate``.
```

Comment (timhoffm) [inline doc/api/next_api_changes/2019-08-12-KS.rst]: Or simply
```suggestion
The parameter ``s`` to `.Axes.annotate` and  `.pyplot.annotate` is renamed to
``text``, matching `.Annotation`.
```

File: lib/matplotlib/tests/test_axes.py [hunk 3]
```diff
@@ -434,6 +434,22 @@ def test_basic_annotate():
                 xytext=(3, 3), textcoords='offset points')
 
 
+def test_annotate_parameter_warn():
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_axes.py]: You don't need all the extra plotting for testing. This should be enough:
~~~
fig, ax = plt.subplots()
with pytest.warns(MatplotlibDeprecationWarning):
    ax.annotate(s='now named text', xy=(0.5, 0.5))
~~~

Also, I recommend to use `pytest.warns(MatplotlibDeprecationWarning, match=...)` to make sure you're catching the right warning.

File: lib/matplotlib/axes/_axes.py [hunk 4]
```diff
@@ -753,9 +753,10 @@ def text(self, x, y, s, fontdict=None, **kwargs):
         self._add_text(t)
         return t
 
+    @cbook._rename_parameter("3.2", "s", "text")
```
Comment (QuLogic) [inline lib/matplotlib/axes/_axes.py]: Next release is 3.3; 3.2 is out already.

Comment (QuLogic) [reply lib/matplotlib/axes/_axes.py]: I guess you noted that earlier, but unless something is broken, I'd rather we leave this deprecation for the next version.

File: doc/api/next_api_changes/2019-08-12-KS.rst [hunk 5]
```diff
@@ -0,0 +1,8 @@
+Changes in parameter names
```
Comment (anntzer) [inline doc/api/next_api_changes/2019-08-12-KS.rst]: this needs to go to the main next_api_changes/behavior.rst (or deprecations.rst?) file.

**Comment by ksunden** [comment]
Rebased on top of master to avoid merge conflict

**Pull request merged**